core/access/service.go

```diff
@@ -0,0 +1,269 @@
+type providerService interface {
+	RevokeAccess(context.Context, domain.Access) error
+}
+
+//go:generate mockery --name=auditLogger --exported --with-expecter
+type auditLogger interface {
+	Log(ctx context.Context, action string, data interface{}) error
+}
+
+//go:generate mockery --name=notifier --exported --with-expecter
+type notifier interface {
+	notifiers.Client
+}
+
+type accessCreation struct {
+	AppealStatus string `validate:"required,eq=active"`
+	AccountID    string `validate:"required"`
+	AccountType  string `validate:"required"`
+	ResourceID   string `validate:"required"`
+}
+
+type Service struct {
+	repo            repository
+	providerService providerService
+
+	notifier    notifier
+	validator   *validator.Validate
+	logger      log.Logger
+	auditLogger auditLogger
+}
+
+type ServiceDeps struct {
+	Repository      repository
+	ProviderService providerService
+
+	Notifier    notifier
+	Validator   *validator.Validate
+	Logger      log.Logger
+	AuditLogger auditLogger
+}
+
+func NewService(deps ServiceDeps) *Service {
+	return &Service{
+		repo:            deps.Repository,
+		providerService: deps.ProviderService,
+
+		notifier:    deps.Notifier,
+		validator:   deps.Validator,
+		logger:      deps.Logger,
+		auditLogger: deps.AuditLogger,
+	}
+}
+
+func (s *Service) List(ctx context.Context, filter domain.ListAccessesFilter) ([]domain.Access, error) {
+	return s.repo.List(ctx, filter)
+}
+
+func (s *Service) GetByID(ctx context.Context, id string) (*domain.Access, error) {
+	if id == "" {
+		return nil, ErrEmptyIDParam
+	}
+	return s.repo.GetByID(ctx, id)
+}
+
+func (s *Service) Prepare(ctx context.Context, appeal domain.Appeal) (*domain.Access, error) {
+	if err := s.validator.Struct(accessCreation{
+		AppealStatus: appeal.Status,
+		AccountID:    appeal.AccountID,
+		AccountType:  appeal.AccountType,
+		ResourceID:   appeal.ResourceID,
+	}); err != nil {
+		return nil, fmt.Errorf("validating appeal: %w", err)
+	}
+
+	return appeal.ToAccess()
+}
+
+func (s *Service) Revoke(ctx context.Context, id, actor, reason string, opts ...Option) (*domain.Access, error) {
+	access, err := s.GetByID(ctx, id)
+	if err != nil {
+		return nil, fmt.Errorf("getting access details: %w", err)
+	}
+
+	revokedAccess := &domain.Access{}
+	*revokedAccess = *access
+	if err := access.Revoke(actor, reason); err != nil {
+		return nil, err
+	}
+	// TODO: remove below logic in future release when appeal no longer for managing access
+	if err := access.Appeal.Revoke(actor, reason); err != nil {
+		return nil, fmt.Errorf("updating appeal status: %s", err)
+	}
+	if err := s.repo.Update(ctx, access); err != nil {
+		return nil, fmt.Errorf("updating access record in db: %w", err)
+	}
+
+	options := s.getOptions(opts...)
+
+	if !options.skipRevokeInProvider {
+		if err := s.providerService.RevokeAccess(ctx, *access); err != nil {
+			if err := s.repo.Update(ctx, access); err != nil {
+				return nil, fmt.Errorf("failed to rollback access status: %w", err)
```

Comment:
what is the impact if we failed to rollback?

Reply:
the access in the guardian and in the actual provider will be not in sync because here we're trying to revoke the provider access and then updating the status in db. This can be solved using transaction